common/lib/xmodule/xmodule/modulestore/search_index.py

```diff
@@ -0,0 +1,213 @@
+""" Code to allow module store to interface with courseware index """
+from __future__ import absolute_import
+
+import logging
+
+from django.utils.translation import ugettext as _
+from opaque_keys.edx.locator import CourseLocator, LibraryLocator
+from search.search_engine_base import SearchEngine
+
+from . import ModuleStoreEnum
+from .exceptions import ItemNotFoundError
+
+# Use default index and document names for now
+
+
+log = logging.getLogger('edx.modulestore')
+
+
+def get_indexer_for_location(locator):
+    if isinstance(locator, CourseLocator):
+        return CoursewareSearchIndexer()
+    elif isinstance(locator, LibraryLocator):
+        return LibrarySearchIndexer()
+    return get_indexer_for_location(locator.course_key)
+
+
+class SearchIndexingError(Exception):
+    """ Indicates some error(s) occured during indexing """
+
+    def __init__(self, message, error_list):
+        super(SearchIndexingError, self).__init__(message)
+        self.error_list = error_list
+
+
+class SearchIndexerBase(object):
+    """
+    Base class to perform XBlock indexing from different modulestores
+    """
+    INDEX_NAME = None
+    DOCUMENT_TYPE = None
+
+    index_on_create = True
+    index_on_update = True
+    index_on_delete = True
+    index_on_publish = True
+
+    def _get_structure_key(self, location):
+        """ Gets structure key from location """
+        return location.course_key
+
+    def _get_location_info(self, structure_key):
+        """ Builds location info dictionary """
+        return {"course": unicode(structure_key)}
+
+    def _fetch_item(self, modulestore, item_location):
+        """ Fetch the item from the modulestore location, log if not found, but continue """
+        raise NotImplemented
+
+    def _id_modifier(self, usage_id):
+        """ Modifies usage_id to submit to index """
+        return usage_id
+
+    def add_to_search_index(self, modulestore, location, delete=False, raise_on_error=False):
+        """
+        Add to courseware search index from given location and its children
+        """
+        error_list = []
+        # TODO - inline for now, need to move this out to a celery task
```

Comment:
@Kelketek this TODO is not mine, I just moved it along with the code. Git diff didn't pick the renaming as the file was extensively modified later, but it's not new code. Original author (Marty James) probably created it.